kyo-stm/shared/src/main/scala/kyo/STM.scala

```diff
@@ -0,0 +1,200 @@
+      * but allows configuring how transaction conflicts are retried.
+      *
+      * @param retrySchedule
+      *   The schedule for retrying failed transactions
+      * @param v
+      *   The transactional computation to run
+      * @return
+      *   The result of the computation if successful
+      */
+    def run[E, A: Flat](retrySchedule: Schedule)(v: A < (STM & Abort[E] & Async))(
+        using frame: Frame
+    ): A < (Async & Abort[E | FailedTransaction]) =
+        useTid {
+            case -1L =>
+                // New transaction without a parent, use regular commit flow
+                Retry[FailedTransaction](retrySchedule) {
+                    withNewTid { tid =>
+                        Var.runWith(Map.empty)(v) { (log, result) =>
+                            IO.Unsafe {
+                                // Attempt to acquire locks and commit the transaction
+                                val (locked, unlocked) =
+                                    // Sort references by identity to prevent deadlocks
+                                    log.toSeq.sortBy((ref, _) => ref.hashCode)
+                                        .span((ref, entry) => ref.lock(entry))
+
+                                if unlocked.nonEmpty then
+                                    // Failed to acquire some locks - rollback and retry
+                                    locked.foreach((ref, entry) => ref.unlock(entry))
+                                    Abort.fail(FailedTransaction(frame))
+                                else
+                                    // Successfully locked all references - commit changes
+                                    locked.foreach((ref, entry) => ref.commit(tid, entry))
+                                    // Release all locks
+                                    locked.foreach((ref, entry) => ref.unlock(entry))
+                                    result
+                                end if
+                            }
+                        }
+                    }
+                }
+            case parent =>
+                // Nested transaction inherits parent's transaction context but isolates RefLog.
+                // On success: changes propagate to parent. On failure: changes are rolled back
+                // without affecting parent's state.
+                val result = Var.isolate.update[RefLog].run(v)
+
+                // Can't return `result` directly since it has a pending STM effect
+                // but it's safe to cast because, if there's a parent transaction,
+                // then there's a frame upper in the stack that will handle the
+                // STM effect in the parent transaction's `run`.
+                result.asInstanceOf[A < (Async & Abort[E | FailedTransaction])]
+        }
+
+    end run
+
+    type RefLog = Map[TRef[Any], Entry[Any]]
```

Comment:
It wasn't trivial because `RefLog` is accessed from multiple source files but I moved it to a proper implementation with an opaque type + extension methods marked as `private[kyo]`.